src/Interpreters/Cache/SLRUFileCachePriority.cpp

```diff
@@ -0,0 +1,285 @@
+    {
+        /// If it is server startup, we put entries in any queue it will fit in,
+        /// but with preference for probationary queue,
+        /// because we do not know the distribution between queues after server restart.
+        if (probationary_queue.canFit(size, lock))
+        {
+            auto lru_iterator = probationary_queue.add(Entry(key_metadata->key, offset, size, key_metadata), lock);
+            return std::make_shared<SLRUIterator>(this, std::move(lru_iterator), false);
+        }
+        else
+        {
+            auto lru_iterator = protected_queue.add(Entry(key_metadata->key, offset, size, key_metadata), lock);
+            return std::make_shared<SLRUIterator>(this, std::move(lru_iterator), true);
+        }
+    }
+    else
+    {
+        auto lru_iterator = probationary_queue.add(Entry(key_metadata->key, offset, size, key_metadata), lock);
+        return std::make_shared<SLRUIterator>(this, std::move(lru_iterator), false);
+    }
+}
+
+bool SLRUFileCachePriority::collectCandidatesForEviction(
+    size_t size,
+    FileCacheReserveStat & stat,
+    EvictionCandidates & res,
+    IFileCachePriority::IteratorPtr reservee,
+    FinalizeEvictionFunc & finalize_eviction_func,
+    const CacheGuard::Lock & lock)
+{
+    /// If `it` is nullptr, then it is the first space reservation attempt
+    /// for a corresponding file segment, so it will be directly put into probationary queue.
+    if (!reservee)
+    {
+        return probationary_queue.collectCandidatesForEviction(size, stat, res, reservee, finalize_eviction_func, lock);
+    }
+
+    /// If `it` not nullptr (e.g. is already in some queue),
+    /// we need to check in which queue (protected/probationary) it currently is
+    /// (in order to know where we need to free space).
+    if (!assert_cast<SLRUIterator *>(reservee.get())->is_protected)
+    {
+        return probationary_queue.collectCandidatesForEviction(size, stat, res, reservee, finalize_eviction_func, lock);
+    }
+
+    /// Entry is in protected queue.
+    /// Check if we have enough space in protected queue to fit a new size of entry.
+    /// `size` is the increment to the current entry.size we want to increase.
+    if (protected_queue.canFit(size, lock))
+        return true;
+
+    /// If not enough space - we need to "downgrade" lowest priority entries from protected
+    /// queue to probationary queue.
+    /// The amount of such "downgraded" entries is equal to the amount
+    /// required to make space for additionary `size` bytes for entry.
+    auto downgrade_candidates = std::make_shared<EvictionCandidates>();
+    FileCacheReserveStat downgrade_stat;
+    FinalizeEvictionFunc noop;
+
+    if (!protected_queue.collectCandidatesForEviction(size, downgrade_stat, *downgrade_candidates, reservee, noop, lock))
+        return false;
+
+    const size_t size_to_downgrade = downgrade_stat.stat.releasable_size;
+
+    if (!probationary_queue.canFit(size_to_downgrade, lock)
```

Comment:
So it will mean that we cannot downgrade N/2 elements from protected to probationary and we have two options:
1. fail space reservation
2. move N/2 elements from protected queue to probationary queue; remove completely other N/2 elements from protected queue.

I decided to go for option 1 because I believed a rule is that elements from protected queue can be removed only by first downgrading to probationary queue and only then they can be completely removed as part of eviction from probationary queue.
But option 2 seems also valid. Do you opt for this way?